### PR TITLE
[Polybench] Update to v4.2.1-beta

### PR DESCRIPTION
**Pull request by jerryzj**
- Old PolyBench/C 3.2 kernels had numerous implementation errors making their outputs to not match what is expected from the mathematical specification of the operation. Many of them did not influence the program behavior. (e.g., the number and type of operations, data
dependences, and overall control-flow was similar to the corrected implementation)

- For more details, please see the README

- Tested on macOS (clang-1500.0.40.1), RHEL (clang 14.0.6 and gcc 13.2.0) and RISCV usermode QEMU rv64gcv (clang upstream-116a1aeff)

- Source: https://sourceforge.net/projects/polybench/files/

### CODE REVIEW

**Comment by fhahn** [comment]
Upgrading to the latest version makes sense to me in general. Could you link to the source in the PR? I guess there's no non-beta version of `v4.2.1`?

**Comment by jerryzj** [comment]
@fhahn I've just updated the link for Polybench, but, unfortunately, there is no non-beta version available.

**Review by rengolin — APPROVED** [review]
This makes sense to me, too. Thanks for the update.

LGTM, but please wait for @fhahn's approval too, to make sure we're not missing anything.

**Comment by fhahn** [comment]
> @fhahn I've just updated the link for Polybench, but, unfortunately, there is no non-beta version available.

Thanks! I am planning to give this a try with top-of-tree Clang tomorrow.

**Comment by jerryzj** [comment]
@fhahn I have just finished testing upstream LLVM (116a1aeff) using RISCV target (rv64gcv/lp64d) under `O3` optimization, and I encountered 2 test failures.

```
test-suite.SingleSource/Benchmarks/Polybench/medley/deriche deriche.test
test-suite.SingleSource/Benchmarks/Polybench/medley/floyd-warshall floyd-warshall.test
```

I don't know if they also failed in your local testing?

**Comment by fhahn** [comment]
> @fhahn I have just finished testing upstream LLVM (116a1aeff) using RISCV target (rv64gcv/lp64d) under `O3` optimization, and I encountered 2 test failures.
> 
> ```
> test-suite.SingleSource/Benchmarks/Polybench/medley/deriche deriche.test
> test-suite.SingleSource/Benchmarks/Polybench/medley/floyd-warshall floyd-warshall.test
> ```
> 
> I don't know if they also failed in your local testing?

All tests are passing on current `main` on ARM64 macOS. @jerryzj do the RISCV tests fail when building with `-O0`? If that's the case, then those new tests may just highlight genuine mis-compiles with RISCV.

**Comment by jerryzj** [comment]
@fhahn yea, these test also failed with `O0`... 😞

File: SingleSource/Benchmarks/Polybench/README
Comment (fhahn) [inline SingleSource/Benchmarks/Polybench/README:84]: It looks like `LICENSE.txt` has not be included? There's `SingleSource/Benchmarks/Polybench//LICENSE`, but that still mentions 3.2 and needs updating I think

**Review by fhahn — COMMENTED** [review]
> @fhahn yea, these test also failed with `O0`... 😞

In a way, it already shows the value of the update! I don't think this needs to be a blocker; once merged, it would be great if you could file issues for the mis-compiles. Having this merged will also make it much easier for people to reproduce the failure.

I left one more inline comment regarding the license files.

File: SingleSource/Benchmarks/Polybench/README
Comment (jerryzj) [reply SingleSource/Benchmarks/Polybench/README:84]: Oops, I missed it. I just updated.

**Comment by jerryzj** [comment]
@fhahn It seems that the tests failed because they hit the maximum output limit. I will try using a smaller problem size for them since the new version defaults to a rather large problem size.

**Comment by jerryzj** [comment]
By default, lit has a constraint for output file size = 100MB

```
TIMEITFLAGS="$TIMEITFLAGS --limit-file-size 104857600"
```

While for deriche, the output file exceeds 100MB...

```
du -sh deriche.out
135M	deriche.out
```

**Comment by jerryzj** [comment]
@fhahn @rengolin The tests are PASS on RISCV now~

**Comment by jerryzj** [comment]
@fhahn If there are no further issues, I feel that we can merge this?

**Review by fhahn — APPROVED** [review]
LGTM, thanks!

**Pull request merged**
